[PATCH v6] Pkt debug

Bill-Fischofer-Linaro opened the pull request:
Restore ODP_ASSERT() calls in packet processing lost in the recent restructure. Also correct a syntax glitch in the previous merge that somehow got past checkpatch.

muvarov commented:
subject has to be "correct syntax". Interesting that separate checkpatch run passed but inside travis it failed.

Bill-Fischofer-Linaro commented:
Changes for v2:

* Change commit subject to avoid checkpatch flagging its own name.

muvarov commented on platform/linux-generic/odp_packet.c:
```diff
@@ -354,6 +354,8 @@ static inline void link_segments(odp_packet_hdr_t *pkt_hdr[], int num)
 			hdr->buf_hdr.seg[i].hdr  = buf_hdr;
 			hdr->buf_hdr.seg[i].data = buf_hdr->base_data;
 			hdr->buf_hdr.seg[i].len  = BASE_LEN;
+			ODP_ASSERT((odp_atomic_store_u32(&hdr->buf_hdr.ref_cnt,
```
Bill, odp_atomic_store_u32() is void. How assert() works here?

Bill-Fischofer-Linaro replied:
It's a compound statement. One of the clever bits of Petri's rework is he doesn't set the initial `ref_cnt` to 1 but rather leaves it as 0 and when the first reference happens bumps it directly to 2. For debugging purposes we want to ensure that the `ref_cnt >= 1` on any free so we need the initial `ref_cnt` to be set to 1, but only for debug builds. The compound statement takes advantage of the fact that in C the "value" of a compound statement is the value of the last component, in this case 1. So the `ODP_ASSERT()` always succeeds and the effect is that the `ref_cnt` is set to 1 only if we're running with `--enable-debug`.

lumag replied:
This is bad, bad, bad idea. Contents of ODP_ASSERT can disappear. So `odp_atomic_store_u32` won't be called.

lumag replied:
@muvarov No, it unrolls to ` ... && (void_func(), 1) ...`. Comma operator.
@Bill-Fischofer-Linaro such asserts need to be reworked.

muvarov replied:
@lumag yep, but what is (void_func(), 1) ? Why gcc compiles that?

Bill-Fischofer-Linaro commented:
C is an expression language. What's being used here is the comma operator. See (https://stackoverflow.com/questions/52550/what-does-the-comma-operator-do)[here] for a discussion of this operator and its uses.

So this expands to `if (... && (void_func(), 1))` which always evaluates to 1, so the `ODP_ASSERT` never fails. But the expression is evaluated (hence the void function is called) if we are in `--enable-debug` mode. 

The compiler must always evaluate all of the comma-separated parts of the expression, which is why you can write things like:

`for (i = 0, j = 0; i < n; i++) ...`

And be confident that both `i` and `j` will be initialized.

The reason for doing this is to avoid having `#ifdefs` in the code to handle debug-only code since `ODP_ASSERT()` does that automatically. It's a common C idiom in preprocessor macros for the same reason.

muvarov commented:
@lumag what did you mean by "@Bill-Fischofer-Linaro such asserts need to be reworked." ? Can I merge it now?

@Bill-Fischofer-Linaro, thanks for explanation. It's really expression and does not matter void or not void it is.

lumag commented:
@muvarov @Bill-Fischofer-Linaro I meant that in my opinion ODP_ASSERT() should contain only conditional code, not functional code (like atomic stores). In fact it would be preferable from my POV to have them under `#ifdef`, if they are really needed only for debugging, with proper comments why. Currently such statements can be easily misunderstood by people reading the code or doing further refactoring.

Bill-Fischofer-Linaro commented:
@lumag I can add some comments to explain what's going on with those "unusual" `ODP_ASSERT()` calls, but we don't like `#ifdef`s in open code. `ODP_ASSERT()` is used to encapsulate code that is only executed if `--enable-debug` is in force.

Bill-Fischofer-Linaro commented:
Changes for v3:

* Rebased to latest master
* Added comments to make it clear what's going on with "unusual" `ODP_ASSERT()` calls and why.

psavol commented on platform/linux-generic/odp_packet.c:
```diff
@@ -354,6 +354,15 @@ static inline void link_segments(odp_packet_hdr_t *pkt_hdr[], int num)
+			 * ODP_ASSERT always succeeds.
+			 */
+			ODP_ASSERT((odp_atomic_store_u32(&hdr->buf_hdr.ref_cnt,
+							 1), 1));
```
Maybe it's better not to change ref_cnt book keeping for debug build. Change of behavior may hide the problem being debugged, etc.

Bill-Fischofer-Linaro replied:
If we want to catch erroneous duplicate packet free calls I don't see how we can avoid that. That's a main source of application errors, especially when references are involved.

The lazy-increment shortcut is a clever efficiency trick for normal processing, but it's counter to what we want in a debug build.

psavol replied:
I'd expect that a debug build adds checks and printf, but does not change e.g. how an algorithm works.

Also store is not enough for double frees, you'd need to inc/dec and check that the value never gets <0.

Bill-Fischofer-Linaro replied:
@psavol The double free guard is the `ODP_ASSERT(ref_cnt > 0)`checks  in `odp_packet_free()` and `odp_packet_free_multi()`. Incrementing and decrementing is already handled by the mainline code. This isn't changing how reference counts are used as the mainline code works the same. It's only removing the lazy increment optimization.

Bill-Fischofer-Linaro commented:
This should be held until we decide if we're going to merge PR #221. If we do I'll resubmit this since it currently references `odp_packet_unshared_len()` in many of the `ODP_ASSERT()` calls.

Bill-Fischofer-Linaro commented:
Changes for v3:

- Rebased to current master

- Removed references to `odp_packet_unshared_len()`

- Switched to use `odp_atomic_fetch_inc_u32()` for `ref_cnt` initialization. This also eliminates the obscure use of the comma operator in `ODP_ASSERT()` processing.

muvarov commented:
I'm ok with v3. Does somebody else wants to add their review? @lumag ?

lumag commented:
@muvarov I stated explicitly, that I do not like functional code inside ODP_ASSERT() arguments. In my opinion this PR should rewritten with `#ifdef's`.

muvarov commented:
@lumag I think Bill removed "," operator. https://github.com/Linaro/odp/pull/180/commits/0768f9527a52768305af75aba4d2ccf300b63705 So it looks like only check for values which is very common to other code.

lumag commented:
@muvarov It still results in code doing refcounting increment under `ODP_ASSERT()` clause. In my opinion this should be better written as
```c
/* Increment counter only in debug case, because ........ */
if (ODP_DEBUG == 1) {
  uint32_t old_cnt = odp_atomic_fetch_inc_u32(&hdr->buf_hdr.ref_cnt);
  ODP_ASSERT(old_cnt == 0); 
}
```

muvarov commented:
@lumag he can not do that. ODP_DEBUG is implementation thing. We can not use it inside tests.

lumag commented:
@muvarov which tests? Commit is changing the `platform/linux-generic/odp_packet.c` file.

Bill-Fischofer-Linaro commented:
Changes in v5:

* Rebased to latest master
* Changed `ODP_ASSERT()` to reflect review comments by @lumag

lumag commented:
@Bill-Fischofer-Linaro Hmm, I don't see any significant changes.

Bill-Fischofer-Linaro commented:
@lumag You asked that the `== 0` not be split across lines. That's the change in v5.

Bill-Fischofer-Linaro commented:
Changes for v6:

* Rebased to current master
* Restructured ref_cnt initialization to avoid side effects in `ODP_ASSERT()` calls.

lumag commented:
Reviewed-by: Dmitry Eremin-Solenikov

muvarov commented:
Merged.
